File: ccc/2010-j4-global-warming.py
#!/usr/bin/env python3
while True:
    # Read input.
    nums = list(map(int, input().split()))
    n = nums[0]
    temperatures = nums[1:]

    # Break out if the first number is 0
    if n == 0: break

    # Calculate changes in temperature
    changes = [b-a for a,b in zip(temperatures[:-1], temperatures[1:])]
    # this also works:
    #changes = [temperatures[i+1] - temperatures[i] for i in range(len(temperatures)-1)]

    # Test each possible cycle length
    for cycle_length in range(1, n):
        # As soon as a proper one is found...
        if all(changes[i] == changes[i-cycle_length] for i in range(cycle_length, len(changes))):
            # print it and break out
            print(cycle_length)
            break

File: ccc/2010/2010-j4-global-warming.cpp
#include <iostream>
using namespace std;

int main ()
{
    int n;
    int temperature[30];
    int change[30];

    for(;;) {
        cin >> n;

        if (n==0) break;
        else if (n==1)
        {
            cout << "0" << endl;
        }

        // Read in temperature values
        for (int i=0; i<n; i++)
        {
            cin >> temperature[i];
        }

        // Calculate changes in temperature (first differences)
        for (int i=0; i<n-1; i++)
        {
            change[i] = temperature[i+1] - temperature[i];
        }

        // Test each possible cycle length (from 1 to n-1)
        for (int cycle_length=1; cycle_length <= n-1; cycle_length++)
        {
            // assume cycle holds...
            bool cycle_holds = true;
            for (int i=cycle_length; i < n-1; i++)
            {
                // ... until we prove that it doesn't
                if (change[i] != change[i-cycle_length])
                {
                    cycle_holds = false;
                    break;
                }
            }

            if (cycle_holds)
            {
                cout << cycle_length << endl;
                break;
            }
        }
    }
}
